datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+    kill signal to the subprocess, if the termination time or kill time has
+    come.
+
+    Parameters
+    ----------
+    cls : type[Protocol]
+        A protocol class that should be extended by the kill-interface
+
+    Returns
+    -------
+    Any
+        If the `cls` is not a subclass of `GeneratorMixIn`, the result of
+        `cls._prepare_result` will be returned.
+
+    Generator
+        If the protocol is a subclass of `GeneratorMixIn`, the `Generator` that
+        is a result of `ThreadedRunner.run()` will be returned.
+    """
+    class KillWrapper(cls):
+        def __init__(self, *args, **kwargs):
+            self.armed = kwargs.pop('armed')
+            self.introduced_timeout = kwargs.pop('introduced_timeout')
+            self.terminate_time = kwargs.pop('terminate_time')
+            kill_time = kwargs.pop('kill_time')
+            self.kill_time = (
+                ((self.terminate_time or 0) + kill_time)
+                if kill_time is not None
+                else kill_time
+            )
+            self.process: subprocess.Popen | None = None
+            self.kill_counter: int = 0
+            super().__init__(*args, **kwargs)
+
+        def arm(self) -> None:
+            self.kill_counter = 0
+            self.armed = True
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed:
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+                self.kill_counter += 1
+
+            # If we set the timeout argument due to a not None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+        cmd: list,
+        protocol_class: type[Protocol],
+        *,
+        cwd: Path | None = None,
+        input: int | IO | bytes | Queue[bytes | None] | None = None,
+        timeout: float | None = None,
+        terminate_time: int | None = None,
+        kill_time: int | None = None,
+) -> Any | Generator:
+
+    introduces_timeout = False
+    if timeout is None:
+        introduces_timeout = True
+        timeout = 1.0
+
+    armed = (
+        False
+        if issubclass(protocol_class, GeneratorMixIn)
+        else True
+    )
```

Comment:
So `armed` means, it is a generated protocol. Why introduce a new term for that?

Reply:
`armed` means that the `KillWrapper`-protocol will start counting timeouts in order to terminate the process. Currently it is set up like this:

- if the protocol is a non-generator-protocol, it (more precisely the `KillWrapper`) will be armed when the process is started, i.e. when the context is entered.
- if the protocol is a generator-protocol, it will be armed when the context is left.

That happens to coincide with `armed == not issubclass(protocol_class, GeneratorMixIn)` (which would be a nice change in the code; will do that), but that is not the only imaginable way to arm the wrapper-class. Since it is an idependent "variable", I kept it separate.